pilz_control/test/integrationtest_pilz_joint_trajectory_controller.py

```diff
@@ -121,155 +120,197 @@ def observeStop(self, max_deceleration = MAX_DECELERATION, timeout = STOP_DURATI
+
+    def _move_to(self, position, duration, expected_error_code=FollowJointTrajectoryResult.SUCCESSFUL):
+        self._start_motion(position=position, duration=duration)
+        return self._wait_for_motion_result(expected_error_code)
+
+    def _start_motion(self, position, duration):
+        self._trajectory_dispatcher.dispatch_trajectory(goal_position=position, time_from_start=duration)
+
+    def _wait_for_motion_result(self, expected_error_code=FollowJointTrajectoryResult.SUCCESSFUL):
+        result = self._trajectory_dispatcher.wait_for_result()
+        return result.error_code == expected_error_code
+
+    def _turn_off_speed_monitoring(self):
+        return self._monitored_cartesian_speed_srv.turn_off_speed_monitoring()
+
+    def _turn_on_speed_monitoring(self):
+        return self._monitored_cartesian_speed_srv.turn_on_speed_monitoring()
+
+    def test_hold_at_startup(self):
+        """Tests that 'hold' mode is active at startup of controller."""
+
+        self.assertTrue(self._move_to(position=DEFAULT_GOAL_POSITION, duration=1,
+                                      expected_error_code=FollowJointTrajectoryResult.INVALID_GOAL),
+                        "Motion did not fail although controller should be in 'hold' mode at startup")
+
+        self.assertTrue(self._hold_srv.request_default_mode(), 'Switch to default mode failed')
+
+        self.assertTrue(self._move_to(position=DEFAULT_GOAL_POSITION, duration=0.5),
+                        "Motion failed although controller should be in 'unhold' mode")
+
+    def test_hold_during_motion(self):
+        """Activate hold mode while robot is moving and evaluate executed stop."""
+        is_executing_srv = IsExecutingServiceWrapper()
+
+        self._start_motion(position = 0.2, duration = 5)
 
         # Wait for movement to commence
-        MovementObserver().observeUntilPositionGreater(0.11, 3)
-        self.assertTrue(is_executing_srv.call(), 'is_executing service returned false.')
-        # Hold robot during movement
-        rospy.loginfo("  - Robot is moving -> Switch into HOLDING mode")
-        self.assertTrue(hold_srv.requestHoldingMode(), 'Switch to Mode HOLDING failed.')
+        MovementObserver().observe_until_position_greater(0.11, 3)
+        self.assertTrue(is_executing_srv.call(), "Controller is not executing")
+        rospy.loginfo("Robot is moving -> Switch into HOLDING mode")
+        self.assertTrue(self._hold_srv.request_holding_mode(), 'Switch to Mode HOLDING failed.')
 
         # Make sure robot stops (not abruptly)
-        MovementObserver().observeStop()
-        rospy.loginfo("  - Stop of robot observed!")
+        MovementObserver().observe_stop()
+        rospy.loginfo("Stop of robot observed!")
+
+        self.assertTrue(self._wait_for_motion_result(), "Motion failed")
+
+        self.assertEqual(GoalStatus.PREEMPTED, self._trajectory_dispatcher.get_last_state(), 'Goal was not preempted.')
+
+    def test_speed_monitoring(self):
```

Comment:
I think this should be extended by moving different joints. Else a robot model with a single continuous joint would suffice.

Reply:
Of course, we can discuss this in common.

Reply:
> I think this should be extended by moving different joints. Else a robot model with a single continuous joint would suffice.

The model has to joints, however, the `RobotMock` class only offers one joint. I tried to come up with a generally usable dummy model and, therefore, added a  second joint. In this case, however, the second joint is not used, therefore, we actually have a robot model with a single continuous joint.

Reply:
Hm. The unit test covers that the `CartesianSpeedMonitor` checks all links of the model that is injected. By repeating it here, we basically would test, that the injection of the robot model is done correctly by the controller.

Reply:
I like the idea of the generally usable dummy model. I think the `RobotMock` should be tailored to it. What do you think?

Reply:
> I like the idea of the generally usable dummy model. I think the `RobotMock` should be tailored to it. What do you think?

OK. I will integrate the second joint into the `RobotMock` class.

Reply:
Done.